tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -2853,6 +2845,13 @@ void CGMSHLSLRuntime::AddConstant(VarDecl *constDecl, HLCBuffer &CB) {
     case hlsl::UnusualAnnotation::UA_RegisterAssignment: {
       if (isGlobalCB) {
         RegisterAssignment *ra = cast<RegisterAssignment>(it);
+        if (ra->RegisterSpace.hasValue()) {
+          DiagnosticsEngine& Diags = CGM.getDiags();
+          unsigned DiagID = Diags.getCustomDiagID(
+            DiagnosticsEngine::Error,
+            "space is only allowed in a constant buffer.");
```

Comment:
I don't think this is quite right.

First, the enclosing context is: we have a constant global variable declaration we wish to add to a constant buffer which may be inside explicit `cbuffer { ... }`, or outside - contributing to the `$Globals` default cbuffer.  `isGlobalCB` is true in the latter case.

I believe this case is triggered when supporting legacy c#, b#, and i# DX9-style register binding, which would define the constant index within the constant array for this variable.  The DX10+ equivalent is the offset within the $Globals constant buffer.  Since DX9 didn't have explicit `cbuffer {}` declarations, it only applies to the $Globals case.  Perhaps the code should at least warn when used inside `cbuffer {}`, since it's not clear that your register annotation will be ignored, but maybe there is an error or warning emitted earlier for this case.

One thing that is clear, space binding shouldn't be included here at all, since it's not a concept that applies within a constant buffer (making the error message incorrect).  In general, I don't think space should be set/allowed for any scenario handled inside this function (though see additional complications below).

There are additional complications for this case that I'm not sure are addressed, but not necessary to address as part of this change, such as:
- we really should only respect the c# bindings here, not b# or i#, since those addressed different constant banks in DX9.
- How does this code handle structures containing numeric types and resources?  Space would apply to resources in a struct in a constant buffer, but only for the resource binding portion.  This is a more widespread problem we need to fix though.
- How about multiple register bindings?

In summary, I think we should generally disallow space here, and figure out which unit tests go through these code paths to see what coverage we have, as well as add some likely missing unit tests.  If, for this PR, you want to get rid of the misleading diagnostic from the special case and add an issue for the follow-up work, I would understand.